src/platform/nrfconnect/FactoryDataParser.h

```diff
@@ -0,0 +1,61 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+struct FactoryDataString
+{
+    void * data;
+    size_t len;
+};
+
+struct FactoryData
```

Comment:
I believe we should 0-initialize all members (here and for factory data string to nullptr) to make the data predictable. This is a very long list of members.

Reply:
It is 0 initialized in the constructor of `FactoryDataProvider` that is creating this structure instance: https://github.com/project-chip/connectedhomeip/pull/19614/files#diff-4521c3129b9a1ff0496c09d5af49e450c8e50b97a538845bcbae1d386f060536R73.

Reply:
The issue I have is that this is C, so I guess "constructor" is out of the question.

Could you maybe create a `inline void FactoryDataInit` to make this obvious (both usage and call sites)?

Reply:
@andy31415 Kamil is absent today, so I tried to address your concerns :)
I added `memset` at the beginning of `ParseFactoryData` so there's no chance uninitialized data is returned after calling the function. Also, I added code to stop booting an application if `ParseFactoryData` returns false and fixed a few other bugs. I removed checking if all mandatory fields are present in `ParseFactoryData` as I'm afraid it's easy to forget adding such a check if a new field is added. Instead, I made sure all getters check if they return initialized data. Let me know if it's better that way.